Source/Core/Common/Arm64Emitter.cpp

```diff
@@ -2003,96 +2004,152 @@ void ARM64XEmitter::ADRP(ARM64Reg Rd, s64 imm)
   EncodeAddressInst(1, Rd, static_cast<s32>(imm >> 12));
 }
 
-// Wrapper around MOVZ+MOVK (and later MOVN)
-void ARM64XEmitter::MOVI2R(ARM64Reg Rd, u64 imm)
+template <typename T, size_t MaxSize>
+class SmallVector final
 {
-  unsigned int parts = Is64Bit(Rd) ? 4 : 2;
-  BitSet32 upload_part(0);
+public:
+  SmallVector() {}
+  SmallVector(size_t size) : m_size(size) {}
 
-  // Always start with a movz! Kills the dependency on the register.
-  bool use_movz = true;
+  void push_back(const T& x) { m_array[m_size++] = x; }
+  void push_back(T&& x) { m_array[m_size++] = std::move(x); }
 
-  if (!imm)
-  {
-    // Zero immediate, just clear the register. EOR is pointless when we have MOVZ, which looks
-    // clearer in disasm too.
-    MOVZ(Rd, 0, ShiftAmount::Shift0);
-    return;
-  }
+  T& operator[](size_t i) { return m_array[i]; }
+  const T& operator[](size_t i) const { return m_array[i]; }
 
-  if ((Is64Bit(Rd) && imm == std::numeric_limits<u64>::max()) ||
-      (!Is64Bit(Rd) && imm == std::numeric_limits<u32>::max()))
-  {
-    // Max unsigned value (or if signed, -1)
-    // Set to ~ZR
-    ARM64Reg ZR = Is64Bit(Rd) ? SP : WSP;
-    ORN(Rd, ZR, ZR, ArithOption(ZR, ShiftType::LSL, 0));
-    return;
-  }
+  size_t size() const { return m_size; }
+  bool empty() const { return m_size == 0; }
 
-  // TODO: Make some more systemic use of MOVN, but this will take care of most cases.
-  // Small negative integer. Use MOVN
-  if (!Is64Bit(Rd) && (imm | 0xFFFF0000) == imm)
-  {
-    MOVN(Rd, ~imm, ShiftAmount::Shift0);
-    return;
-  }
+private:
+  std::array<T, MaxSize> m_array;
+  size_t m_size = 0;
+};
 
-  // XXX: Use MOVN when possible.
-  // XXX: Optimize more
-  // XXX: Support rotating immediates to save instructions
-  for (unsigned int i = 0; i < parts; ++i)
+template <typename T>
+void ARM64XEmitter::MOVI2RImpl(ARM64Reg Rd, T imm)
+{
+  enum class Approach
   {
-    if ((imm >> (i * 16)) & 0xFFFF)
-      upload_part[i] = 1;
-  }
+    MOVZ,
+    MOVN,
+    ADR,
+    ADRP,
+  };
 
-  u64 aligned_pc = (u64)GetCodePtr() & ~0xFFF;
-  s64 aligned_offset = (s64)imm - (s64)aligned_pc;
-  // The offset for ADR/ADRP is an s32, so make sure it can be represented in that
-  if (upload_part.Count() > 1 && std::abs(aligned_offset) < 0x7FFFFFFFLL)
+  struct Part
   {
-    // Immediate we are loading is within 4GB of our aligned range
-    // Most likely a address that we can load in one or two instructions
-    if (!(std::abs(aligned_offset) & 0xFFF))
+    Part() = default;
+    Part(u16 imm_, ShiftAmount shift_) : imm(imm_), shift(shift_) {}
+
+    u16 imm;
+    ShiftAmount shift;
+  };
+
+  constexpr size_t max_parts = sizeof(T) / 2;
+
+  // Have best_parts initially be big enough to ensure try_base overwrites it
+  SmallVector<Part, max_parts> best_parts(max_parts + 1);
```

Comment:
Why max_parts + 1? Wouldn't that be larger than the vector's capacity (max_parts)?

Reply:
Oh I see, it's for the instructions_required check... This smells a bit fishy though -- could we avoid out of bounds indices/sizes? :)

Reply:
This is the most efficient way to implement it as far as I can tell, but yeah, it does look rather fishy even though it does work correctly. I suppose the options would be to either make the compile-time size one bigger or to have some kind of variable `bool must_overwrite_best_parts` which is set to true only the first time. Which would you prefer?

Reply:
must_overwrite_best_parts sounds good!